libcontainer/system/linux.go

```diff
@@ -102,3 +106,40 @@ func GetSubreaper() (int, error) {
 
 	return int(i), nil
 }
+
+func ExecutableMemfd(comment string, flags int) (*os.File, error) {
+	// Try to use MFD_EXEC first. On pre-6.3 kernels we get -EINVAL for this
+	// flag. On post-6.3 kernels, with vm.memfd_noexec=1 this ensures we get an
+	// executable memfd. For vm.memfd_noexec=2 this is a bit more complicated.
+	// The original vm.memfd_noexec=2 implementation incorrectly silently
+	// allowed MFD_EXEC[1] -- this should be fixed in 6.6. On 6.6 and newer
+	// kernels, we will get -EACCES if we try to use MFD_EXEC with
+	// vm.memfd_noexec=2 (for 6.3-6.5, -EINVAL was the intended return value).
+	//
+	// The upshot is we only need to retry without MFD_EXEC on -EINVAL because
+	// it just so happens that passing MFD_EXEC bypasses vm.memfd_noexec=2 on
+	// kernels where -EINVAL is actually a security denial.
+	memfd, err := unix.MemfdCreate(comment, flags|unix.MFD_EXEC)
+	if errors.Is(err, unix.EINVAL) {
+		memfd, err = unix.MemfdCreate(comment, flags)
+	}
+	if err != nil {
+		if errors.Is(err, unix.EACCES) {
+			logrus.Infof("memfd_create(MFD_EXEC) failed, possibly due to vm.memfd_noexec=2 -- falling back to less secure O_TMPFILE")
+		}
+		err := os.NewSyscallError("memfd_create", err)
+		return nil, fmt.Errorf("failed to create executable memfd: %w", err)
+	}
+	return os.NewFile(uintptr(memfd), "/memfd:"+comment), nil
+}
+
+// Copy is a wrapper around io.Copy that continues the copy despite EINTR.
+func Copy(dst io.Writer, src io.Reader) (written int64, err error) {
+	for {
+		n, err := io.Copy(dst, src)
+		written += n
+		if !errors.Is(err, unix.EINTR) {
+			return written, err
+		}
+	}
+}
```

Comment:
AFAIK go now handles EINTR properly, so there is no need for this wrapper.

Reply:
Ah, they made all fd-related syscalls use EINTR wrappers in https://go-review.googlesource.com/c/go/+/249178 it seems. They didn't mention that in the Go 1.16 release notes, it seems...